dbms/src/Flash/Mpp/tests/gtest_mpptunnel.cpp

```diff
@@ -671,5 +570,74 @@ TEST_F(TestMPPTunnel, AsyncWriteError)
     }
 }
 
+/// Test Local MPPTunnel
+TEST_F(TestMPPTunnel, LocalConnectWriteDone)
+try
+{
+    const size_t tunnel_num = 3;
+    size_t send_data_packet_num = 3;
+    auto [receiver, tunnels] = prepareLocal(tunnel_num);
+
+    std::vector<std::thread> threads;
+    for (size_t i = 0; i < tunnels.size(); ++i)
+    {
+        auto run_tunnel = [sender_tunnel = tunnels[i]->getLocalTunnelSender(), send_data_packet_num]() {
+            for (size_t i = 0; i < send_data_packet_num; ++i)
+                sender_tunnel->push(newDataPacket("111"));
+            sender_tunnel->finish();
+        };
+        threads.push_back(std::thread(tunnelRun<decltype(run_tunnel)>, std::move(run_tunnel)));
+    }
+
+    receiver->run();
+
+    for (auto & thread : threads)
+        thread.join();
+
+    EXPECT_EQ(receiver->getReceivedMsgs().size(), tunnel_num * send_data_packet_num);
+}
+CATCH
+
+TEST_F(TestMPPTunnel, LocalConnectWriteCancel)
+try
+{
+    const size_t tunnel_num = 1;
+    auto [receiver, tunnels] = prepareLocal(tunnel_num);
+
+    std::thread t(&MockExchangeReceiver::run, receiver.get());
+
+    tunnels[0]->write(newDataPacket("First"));
+    tunnels[0]->write(newDataPacket("Second"));
+    tunnels[0]->close("Cancel", true);
+    GTEST_ASSERT_EQ(getTunnelFinishedFlag(tunnels[0]), true);
+
+    t.join();
+    auto result_size = receiver->getReceivedMsgs().size();
+    GTEST_ASSERT_EQ(result_size >= 0 && result_size <= 2, true);
+    GTEST_ASSERT_EQ(receiver->getErrMsg(), "Cancel");
+}
+CATCH
+
+TEST_F(TestMPPTunnel, LocalConsumerFinish)
+try
+{
+    const size_t tunnel_num = 1;
+    auto [receiver, tunnels] = prepareLocal(tunnel_num);
+
+    std::thread t(&MockExchangeReceiver::run, receiver.get());
+
+    tunnels[0]->write(newDataPacket("First"));
+    tunnels[0]->write(newDataPacket("Second"));
+    tunnels[0]->getLocalTunnelSender()->consumerFinish("");
+    GTEST_ASSERT_EQ(getTunnelSenderConsumerFinishedFlag(tunnels[0]->getTunnelSender()), true);
+
+    t.join();
+    auto result_size = receiver->getReceivedMsgs().size();
+    GTEST_ASSERT_EQ(result_size == 2, true);
+    GTEST_ASSERT_EQ(receiver->getReceivedMsgs()[0]->packet->getPacket().data(), "First");
+    GTEST_ASSERT_EQ(receiver->getReceivedMsgs()[1]->packet->getPacket().data(), "Second");
+}
+CATCH
+
```

Comment:
Is it possible to migrate the other previous unit tests?
`LocalConnectWhenFinished`, `LocalConnectWhenConnected` and etc

Reply:
> Is it possible to migrate the other previous unit tests? `LocalConnectWhenFinished`, `LocalConnectWhenConnected` and etc

Done